### PR TITLE
feat: domain typing

### PR DESCRIPTION
**Pull request by GregoryAlbouy**
## Description

Provide domain typings for Benchttp: input config, output report and typing constraints regarding tests input

### CODE REVIEW

File: src/benchttp/metrics.ts
Comment (moreirathomas) [inline src/benchttp/metrics.ts:14]: `MetricId` circularly references itself through MetricId` -> `MetricField` -> `DurationField | NumberField` -> `SingleMetricField<MetricId>`.

I suggest to simply define `MetricId` as the union of all ids.

```suggestion
type MetricId = DurationMetricId | NumberMetricId
```

File: src/benchttp/run.ts
Comment (moreirathomas) [inline src/benchttp/run.ts:40]: Suggestion: `Stats` could be enough as there is no reference to time data here, simply numbers.

File: src/benchttp/metrics.ts
Comment (GregoryAlbouy) [reply src/benchttp/metrics.ts:14]: I noticed that, but TypeScript seems to handle that pretty well.
Let's change it back to the union, I agree it can be confusing.

Comment (moreirathomas) [reply src/benchttp/metrics.ts:14]: Maintenance would be small as an error would pop quickly if they get out of sync.
My main point against it is the circumvoluted way of defining the types.

I let you choose what you prefer ! :)

**Review by moreirathomas — APPROVED** [review]
I would simply change the title of the pr to something other than "docs" as there are functional changes.

Other than that we're good to go 🙂

File: src/benchttp/metrics.ts
Comment (GregoryAlbouy) [reply src/benchttp/metrics.ts:14]: I'll let @alex-mngt decide then 😛

Comment (GregoryAlbouy) [reply src/benchttp/metrics.ts:14]: Finally decided to move on with the current `MetricField['id']`, we'll have the occasion to discuss it again because these definitions will probably change in a near future (see https://github.com/benchttp/engine/pull/41).
Merging anyway, so we can benefit from this types without being blocked by the linked PR.

**Pull request merged**
